### DISCUSSION TITLE
86box as device server for qemu?

### DISCUSSION THREAD

**Discussion by Randrianasulu** [86Box/86Box#4456] [Ideas]
Well, I know it will sounds crazy, especially in the light of last comment at https://github.com/86Box/86Box/discussions/3670 (device model / timings in 86box tied to its own cpu)  but at least theoretically what kind of protocol should be involved if anyone ever tried it?

What qemu should export, assuming 86box run in its own process and load or more accurately expose (proprietary) roms all by itself? With result (screen buffer, sound buffer, network buffer...) transfered back to qemu ...

Some sort of virtual pci bus, for graphics (always hardest part), I guess?

Reasons:
All this retyping so time consuming no one likely to do it for full reimplementation of 86box devices on qemu side

Legal reasons about proprietary roms.

PCI devices (at least non-itegrated ones) are naturally on extension card physically and on pci bus logically. There is quite big struggle  toward understanding how exactly say Ati's r128/radeon acts on lowest subcomponent level (CCE/command processor with proprietary microcode), so I thought may be "exposing" some mach64 or voodoo3 device to qemu-system-ppc/sparc might be interesting for more complete emulation (@TC1995 )

Cooperation is hard, so ..... lets try it? Developer I tend to bother rn (and since some years already - was tester for his r128 qemy patches):

https://osdn.net/users/balaton/

**Comment by DDXofficial** [comment]
This was a bit complex to understand at first, I'm afraid. Are you proposing some kind of solution/method for 86Box and QEMU to work in tandem so that 86Box emulates only parts of the computer and QEMU handles the rest?

**Comment by Randrianasulu** [comment]
Well, yes? Problem is, I do not know enough of qemu internals  for asserting it will work if enough effort will be put into it.

So, I am trying to get 86box side provide some explanations about what kind of interface might work for this use-case.

In "discussions" because it really not clear for me what kind of requrements such use case might impose on both codevases, so hopefully there will be , eh, discussion about how many obstacles need to be defeated for such idea to materialize even in proof of concept form. I am definitely inspired by threads like
"What need be done for NeXT emulator" that after few years of "just talk" attracted developer who was finally able to make such emulator, Previous, and keep improving it ever since.

**Comment by Randrianasulu** [comment]
for example, qemu already have virtio-gpu-host for 3d rendering in separate process.

 https://patchwork.ozlabs.org/project/qemu-devel/cover/20180713130916.4153-1-marcandre.lureau@redhat.com/

but it seems to aim at specific virtio protocol, and spec for virtio is HUGE (for me, at least).

https://wiki.osdev.org/Virtio

Also, this is not obvious for me if protocol allows to send result of processing  back to qemu? In theory I can imagine 86box acting as terminal (display, keyb, mouse) to qemu, but I am not sure how technically possible it is.
